trieste/models/gpflow/models.py

```diff
@@ -1362,3 +1368,259 @@ def covariance_between_points(
+        """
+        fidelity_query_points = convert_query_points_for_fidelity(
+            dataset.query_points, fidelity - 1
+        )
+        residuals = (
+            dataset.observations - self.rho[fidelity] * self.predict(fidelity_query_points)[0]
+        )
+        return residuals
+
+    def sample(self, query_points: TensorType, num_samples: int) -> TensorType:
+        """
+        Sample `num_samples` samples from the posterior distribution at `query_points`
+
+        :param query_points: The query points at which to sample of shape [N, D+1], where the
+            final column of the final dimension contains the fidelity of the query point
+        :param num_samples: The number of samples (S) to generate for each query point.
+        :return: samples from the posterior of shape [S, N, P]
+        """
+        (
+            query_points_wo_fidelity,
+            query_points_fidelity_col,
+        ) = check_and_extract_fidelity_query_points(
+            query_points, max_fidelity=self.num_fidelities - 1
+        )
+
+        signal_sample = self.lowest_fidelity_signal_model.sample(
+            query_points_wo_fidelity, num_samples
+        )
+
+        for fidelity in range(self.num_fidelities):
+            if fidelity == 0:
+                fidelity_residual_sample = 0
+            else:
+                fidelity_residual_sample = self.fidelity_residual_models[fidelity].sample(
+                    query_points_wo_fidelity, num_samples
+                )
+
+            new_fidelity_signal_sample = (
+                self.rho[fidelity] * signal_sample + fidelity_residual_sample
+            )
+
+            mask = query_points_fidelity_col >= fidelity
+
+            signal_sample = tf.where(mask, new_fidelity_signal_sample, signal_sample)
+
+        return signal_sample
+
+    def predict_y(self, query_points: TensorType) -> tuple[TensorType, TensorType]:
+        """
+        Predict the marginal mean and variance at `query_points` including observation noise
+
+        :param query_points: Query points with shape [N, D+1], where the
+            final column of the final dimension contains the fidelity of the query point
+        :return: mean: The mean at query_points with shape [N, P],
+                 and var: Th variance at query_points with shape [N, P]
+        """
+
+        f_mean, f_var = self.predict(query_points)
+        query_points_fidelity_col = query_points[:, -1:]
+
+        # Get fidelity 0 observation noise
+        observation_noise = (
+            tf.ones_like(query_points_fidelity_col)
+            * self.lowest_fidelity_signal_model.get_observation_noise()
+        )
+
+        for fidelity in range(1, self.num_fidelities):
+
+            fidelity_observation_noise = (
+                self.rho[fidelity] ** 2
+            ) * observation_noise + self.fidelity_residual_models[fidelity].get_observation_noise()
+
+            mask = query_points_fidelity_col >= fidelity
```

Comment:
why `>=` and not equal?

Reply:
Because we need to continually feed predictions from lower fidelities through higher fidelities to get high fidelity predictions